lib/src/element_type.dart

```diff
@@ -115,25 +115,26 @@ class UndefinedElementType extends ElementType {
 
   @override
   String get name {
-    if (type.isDynamic) {
-      if (returnedFrom != null &&
-          (returnedFrom is DefinedElementType &&
-              (returnedFrom as DefinedElementType).element.isAsynchronous)) {
-        return 'Future';
-      } else {
-        return 'dynamic';
-      }
-    }
+    if (isImpliedFuture) return 'Future';
     if (type.isVoid) return 'void';
-    if (type.isBottom) return 'Never';
-    assert(false,
+    assert({'Never', 'void', 'dynamic'}.contains(type.element.name),
         'Unrecognized type for UndefinedElementType: ${type.toString()}');
-    return '';
+    return type.element.name;
   }
 
+  /// Returns true if this type is an implied `Future`.
+  bool get isImpliedFuture => (type.isDynamic &&
```

Comment:
This concept is wild. Since it is the existing behavior, I'm ok with it. But, wild :)

Reply:
Indeed.  At least now it is explicit and it should be clear what's happening here, even if it isn't clear whether we should be doing it at all :-D